src/main/java/edu/kit/ipd/crowdcontrol/objectservice/moneytransfer/MoneyTransferManager.java

```diff
@@ -358,11 +376,65 @@ private String loadMessage(String path) throws MoneyTransferException {
+            throw new MoneyTransferException("The file at \"" + path + "\" couldn't be read. Please secure, that the file isn't corrupt", e);
         }
 
         return content.toString();
     }
+
+
+    protected static BigDecimal getExchangeRate(String sourceCurrency, String destinationCurrency) throws MoneyTransferException {
+        LOGGER.trace("Started fetching currency exchange rates from " + sourceCurrency + " to " + destinationCurrency + ".");
+
+        CloseableHttpClient httpclient = HttpClientBuilder.create().build();
+        HttpGet httpGet = new HttpGet("http://quote.yahoo.com/d/quotes.csv?s=" + sourceCurrency + destinationCurrency + "=X&f=l1&e=.csv");
+        ResponseHandler<String> responseHandler = new BasicResponseHandler();
+
+        String responseBody;
+        try {
+            responseBody = httpclient.execute(httpGet, responseHandler);
+            httpclient.close();
+        } catch (IOException e) {
+            LOGGER.error("Fetching currency exchange rates failed.");
+            throw new MoneyTransferException("There was an error while getting exchange rates from " + sourceCurrency + " to " + destinationCurrency + ".", e);
+        }
+
+        BigDecimal rate = new BigDecimal(responseBody.substring(0, responseBody.length() - 2));
+
+        boolean isEurUsdConversion = false;
+        if ((sourceCurrency.equals("EUR") && destinationCurrency.equals("USD")) || (sourceCurrency.equals("USD") && destinationCurrency.equals("EUR"))) {
+            isEurUsdConversion = true;
+        }
+
+        if (rate.compareTo(new BigDecimal(2)) == 1 && isEurUsdConversion) {
+            throw new MoneyTransferException("Exchange rate in an EUR/USD conversion is above 2. The exchange rate must not be above 2.");
+        } else if (rate.compareTo(new BigDecimal("0.5")) == -1 && isEurUsdConversion) {
+            throw new MoneyTransferException("Exchange rate in an EUR/USD conversion is below 0.5. The exchange rate must not be below 0.5.");
+        }
+
+        LOGGER.trace("Completed fetching currency exchange rates.");
+
+        return rate;
+    }
+
+    private int exchangeTo(int amount, int sourceCurrencyCode, int destCurrencyCode) throws MoneyTransferException {
+        CurrencyUnit sourceCurr = CurrencyUnit.ofNumericCode(sourceCurrencyCode);
+        CurrencyUnit destCurr = CurrencyUnit.ofNumericCode(destCurrencyCode);
+
+        Money money = Money.zero(sourceCurr);
+        money = money.plusMinor(amount);
+        BigDecimal exchangeRate;
+        String key = Integer.toString(sourceCurrencyCode) + Integer.toString(destCurrencyCode);
+        if (exchangeRates.containsKey(key)) {
+            exchangeRate = exchangeRates.get(key);
+        } else {
+            exchangeRate = getExchangeRate(sourceCurr.getCurrencyCode(), destCurr.getCurrencyCode());
+            exchangeRates.put(key, exchangeRate);
```

Comment:
oh no, there is a bug: it gets resetted at the end. And when an exception occurs, but some exchange rates got already fetched, the map doesn't get resetted. I'll fix it, and then merge it please ;)